nanocoat/lib/scritchui/scritchPencilRasterShapes.c

```diff
@@ -17,7 +17,399 @@
+		if (filledPixels == NULL)
+		{
+			error = sjme_error_outOfMemory(NULL, (w * h) *
+				sizeof(sjme_jboolean));
+			goto fail_any;
+		}
+		memset(filledPixels, 0, (w * h) * sizeof(sjme_jboolean));
+	}
+
+	/**
+	 * This is just drawArc's Bresenham midpoint circle algorithm modified to 
+	 * draw arcs from the center to the edge and check if a pixel was already
+	 * painted before. Works great but is slow on larger transparent arcs.
+	 * Trying to use a scanline approach could result in better performance due
+	 * to the simpler checks (just go from yMax to yMin while making sure a y 
+	 * position is never used twice) but doesn't seem to work right in some
+	 * edge cases (negative angles, too wide/narrow arc angles, etc) so this
+	 * one, albeit slow, is what we'll be going with at the moment.
+	 * 
+	 * At least it has the upside of generating far more stable arcs that fully
+	 * match their outline compared to Java AWT's algorithm.
+	 * 
+	 * TODO: Optimize this later
+	 */ 
+	
+	centerX = x + w / 2.0f;
+	centerY = y + h / 2.0f;
+	radiusX = w / 2.0f;
+	radiusY = h / 2.0f;
+	startAngleRad = startAngle * 0.017453292f;
+	endAngleRad = ((startAngle + arcAngle) * 0.017453292f) 
+	- startAngleRad;
+
+	maxRad = radiusX > radiusY ? radiusX : radiusY;
+
+	steps = fabs(arcAngle * ((w + h) / 2.0f) / 50.0f);
+	innerX, innerY;
+
+	for(int i = 0; i < steps; i++) 
+	{
+		angle = startAngleRad + (i * endAngleRad / steps);
+
+		for(int j = 0; j < maxRad; j++) 
+		{
+			innerX = round(centerX + radiusX * cos(angle) * (j / maxRad));
+			innerY = round(centerY + radiusY * sin(angle) * (j / maxRad));
+
+			/*
+			 * Make sure we're not drawing out of bounds. Or accessing the
+			 * alpha buffer at an invalid position with innerX-x or innerY-y
+			 */
+			if(innerX < clipLine->s.x || innerX >= clipLine->e.x
+				|| innerY < clipLine->s.y || innerY >= clipLine->e.y 
+				|| innerX-x < 0 || innerY-y < 0)
+				continue;
+
+			/** 
+			 * Only draw if opaque, or if the alpha buffer is not yet filled
+			 * for the current position.
+			 */
+			if(hasAlpha ? !*(filledPixels + ((innerY-y) * w + innerX-x)) : 1)
+			{
+				if(hasAlpha)
+					*(filledPixels + ((innerY-y) * w + innerX-x)) = 1;
+				error |= drawPixel(g, innerX, innerY);
+			}
+		}
+	}
+
+	if(hasAlpha)
+		sjme_alloca_free(filledPixels);
+
+	/* Failed? */
+	if (sjme_error_is(error))
+		goto fail_any;
+	
+	/* Success? */
+	return error;
+	
+fail_any:
+	
+	return sjme_error_default(error);
+}
+
+sjme_errorCode sjme_scritchpen_corePrim_fillPolygon(
+	sjme_attrInNotNull sjme_scritchui_pencil g,
+	sjme_attrInValue sjme_jint xPoints[],
+	sjme_attrInPositive sjme_jint xOffset,
+	sjme_attrInValue sjme_jint yPoints[],
+	sjme_attrInPositive sjme_jint yOffset,
+	sjme_attrInPositive sjme_jint nPoints)
+{
+	sjme_errorCode error;
+	sjme_scritchui_pencilDrawHorizFunc drawHoriz;
+	int ymin = 2147483647, ymax = -2147483648, intersectionCount = 0, xStart,
+	xEnd, temp, dy, j;
+	int intersections[nPoints];
+	long ix;
```

Comment:
`sjme_jint` for `int` and `sjme_jlong` for `long` respectfully.

The initialization should be moved out as initializer mixing is not possible with all compilers.

Additionally, there is `INT32_MAX` and `INT32_MIN` for those magic numbers.

Reply:
The jlongNative variable was made into a sjme_jint as well. The long was mostly an "additional safety measure" i put in place when testing edge cases in FreeJ2ME-Plus, the likelihood of the arithmetic here going out of range of a standard int are microscopic, if not nonexistent.